movies-manager/src/containers/main/main.tsx

```diff
@@ -161,35 +148,65 @@ export function Main(props: IMainProps): JSX.Element {
             </li>;
         });
 
-    return <main className={blockName}>
+    return loadConfig.isLoading ? <LoadingIndicator /> :
+        !loadConfig.isLoaded && <ErrorHandler errorMessage={errorInfo.message}/>
+        || <main className={blockName}>
         <Modal isOpen={isFormDialogOpen} handleClose={hideModal}>
-            <FormPage onSaveChanges={updateMoviesSet} movie={ movieToEdit }/>
+            <FormPageWithState onSaveChanges={updateMoviesSet} movie={ movieToEdit }/>
         </Modal>
         <Modal isOpen={isDeleteDialogOpen} handleClose={hideModal}>
-            <DeleteModal  onDeleteConfirm={deleteMovie} title={movieToEdit.title}/>
+            <DeleteModal onDeleteConfirm={onDeleteMovie} title={movieToEdit.title}/>
         </Modal>
         <Wrapper>
-            { props.areDetailsVisible && movieWithDetails ?
-                <Details { ...movieWithDetails }/> : <Search/> }
+            { props.areDetailsVisible ? <DetailsWithState /> : <Search onSearchClick={searchMovies} /> }
         </Wrapper>
         <div className={`${blockName}__separator`} />
         <Wrapper>
             <>
                 <section className={`${blockName}__filter`}>
                     <ResultFilter
-                        onGenreClick={setCurrentGenre}
-                        { ...moviesGenresConfig }/>
+                        onGenreClick={sortMoviesByGenre}
+                        { ...storeGenresConfig }/>
                     <ResultSort
-                        onSortClick={updateMoviesSortConfig}
+                        onSortClick={sortMoviesByField}
+                        showOptionList={isSortByOpen}
                         {...moviesSortConfig}/>
                 </section>
                 <div className={`${blockName}__amount`}>
-                    <strong className='strong'>{movies.length}</strong> movies found
+                    <strong className='strong'>{moviesStore.length}</strong> movies found
                 </div>
                 <ul className={`${blockName}__cards-list`}>
                     {moviesCards}
                 </ul>
             </>
         </Wrapper>
     </main>;
-}
\ No newline at end of file
+}
+
+const mapStateToProps = (state: IStoreState, ownProps: IMainProps) => {
+    return {
+      ...ownProps,
+      moviesStore: state.moviesConfig.movies,
+      loadConfig: {
+          isLoaded: state.moviesConfig.isLoaded,
+          isLoading: state.moviesConfig.isLoading,
+      },
+      errorInfo: state.moviesConfig.error,
+      storeGenresConfig: state.filters.genresConfig,
+      moviesSortConfig: state.filters.sortByConfig,
+    };
+};
+
+const dispatchToProps = ((dispatch: any) => {
+    return {
+        loadData: (params: TNullable<IQueryParams>) => { dispatch(getMoviesFromServer(params)); },
```

Comment:
loadData should be in Search component

Reply:
disagree, because:
1. search component is dumb as per my implementation
2. when I set any movie as detail, if I place loading logic in search component - I will not be able to see relevant movies, when I'll click genre tab or sortBy

Reply:
Component with input and button is smart by default.

Reply:
Agry and in my case it is component initiating change is Main, not? In such a case if I place loading in search component I should to pass loading also from sortBy and genres tab, not?

Reply:
Why input should be smart? It event do not have to keep state